test/cpp/api/any.cpp

```diff
@@ -0,0 +1,200 @@
+    }
+    SECTION("pointers") {
+      std::string s("hello");
+      std::string* p = &s;
+      auto any = make_any(p);
+      REQUIRE(any.try_get<std::string*>() != nullptr);
+      REQUIRE(*any.get<std::string*>() == "hello");
+    }
+    SECTION("references") {
+      std::string s("hello");
+      const std::string& t = s;
+      auto any = make_any(t);
+      REQUIRE(any.try_get<std::string>() != nullptr);
+      REQUIRE(any.get<std::string>() == "hello");
+    }
+  }
+  SECTION("move is allowed") {
+    auto any = make_any(5);
+    SECTION("construction") {
+      auto copy = make_any(std::move(any));
+      REQUIRE(copy.try_get<int>() != nullptr);
+      REQUIRE(copy.get<int>() == 5);
+    }
+    SECTION("assignment") {
+      auto copy = make_any(10);
+      copy = std::move(any);
+      REQUIRE(copy.try_get<int>() != nullptr);
+      REQUIRE(copy.get<int>() == 5);
+    }
+  }
+  SECTION("type_info is correct") {
+    SECTION("int") {
+      auto any = make_any(5);
+      REQUIRE(any.type_info().hash_code() == typeid(int).hash_code());
+    }
+    SECTION("const char") {
+      auto any = make_any("hello");
+      REQUIRE(any.type_info().hash_code() == typeid(const char*).hash_code());
+    }
+    SECTION("std::string") {
+      auto any = make_any(std::string("hello"));
+      REQUIRE(any.type_info().hash_code() == typeid(std::string).hash_code());
+    }
+  }
+}
+
+TEST_CASE("any-module") {
+  SECTION("int(int)") {
+    struct M : nn::Module {
+      int forward(int x) {
+        return x;
+      }
+    };
+    AnyModule any(M{});
+    REQUIRE(any.forward(5).get<int>() == 5);
+  }
+  SECTION("const char*(const char*)") {
+    struct M : nn::Module {
+      const char* forward(const char* x) {
+        return x;
+      }
+    };
+    AnyModule any(M{});
+    REQUIRE(any.forward("hello").get<const char*>() == std::string("hello"));
+  }
+
+  SECTION("string(int, const double)") {
+    struct M : nn::Module {
+      std::string forward(int x, const double f) {
+        return std::to_string(static_cast<int>(x + f));
+      }
+    };
+    AnyModule any(M{});
+    int x = 4;
+    REQUIRE(any.forward(x, 3.14).get<std::string>() == std::string("7"));
+  }
+
+  SECTION("Variable(string, const string&, string&&)") {
+    struct M : nn::Module {
+      autograd::Variable
+      forward(std::string a, const std::string& b, std::string&& c) {
+        const auto s = a + b + c;
+        return autograd::make_variable(
+            at::ones(at::CPU(at::kFloat), {static_cast<int64_t>(s.size())}));
+      }
+    };
+    AnyModule any(M{});
+    REQUIRE(
+        any.forward(std::string("a"), std::string("ab"), std::string("abc"))
+            .get<autograd::Variable>()
+            .sum()
+            .toCInt() == 6);
+  }
+  SECTION("error cases") {
+    struct M : nn::Module {
+      int forward(float x) {
+        return x;
+      }
+    };
+    AnyModule any(M{});
+    REQUIRE_THROWS_WITH(
+        any.forward(5.0),
```

Comment:
I guess this test suggests no implicit conversions happen. That's annoying.

Reply:
No! Sad. We could add manual stuff to do implicit conversions for primitive types? Should we?

Reply:
I think implicit conversions would be cool. Another quality of life thing is to make a default `.get()` return an `at::Tensor`, since in a lot of cases you get 1-var input and outputs.